### PR TITLE
Remove null from startup-dependencies.json while writing and reading …

### PR DESCRIPTION
**Pull request by li-gsalia**
…the file.

A couple of users complained about not being able to execute flows. Debugging the
issue revealed that thinZip was enabled for this project and startup dependencies
that are saved in the DB contains a 'null'.

When 'null' is encountered in the dependencies json stored in the DB, an error occurs
while created Dependency object from the json since 'null' is not expected.

This change adds filters and checks to remove 'null' while writing and reading startup
dependencies. Also added unit test to ensure the code checks for null while writing
startup-dependencies.json to disk and while reading the dependencies json object
from the DB.

### CODE REVIEW

File: test/src/main/java/azkaban/test/executions/ThinArchiveTestUtils.java [hunk 1]
```diff
@@ -120,6 +130,22 @@ public static String getRawJSONDepsABC() {
         "}";
   }
 
+  /**
+   * Create a string of valid raw json that adds a null to the list of valid dependencies.
+   *
+   * @return raw json string containing valid dependencies and a null.
+   */
+  public static String getRawJSONDepsNullAndABC() {
```
Comment (jakhani) [inline test/src/main/java/azkaban/test/executions/ThinArchiveTestUtils.java:138]: Why the method name has ABC? Can you please share what does A B and C denotes?

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java [hunk 2]
```diff
@@ -173,4 +178,18 @@ public static void validateDependencyHash(final File f, final Dependency d)
...
   }
+
+  /**
+   * Helper method to filer out null from the set containing Dependency objects.
+   * Eg: If input set: {null, obj1, obj2 }, then output set: {obj1, obj2}
+   *
+   * @param dependencies Set of Dependency objects from which null has to be filtered out.
+   * @return Set of Dependency without any null in it.
+   */
+  public static Set<Dependency> filterNullFromDeps(Set<azkaban.spi.Dependency> dependencies) {
```
Comment (jakhani) [inline azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:189]: Can you please import azkaban.spi package?

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java [hunk 3]
```diff
@@ -173,4 +178,18 @@ public static void validateDependencyHash(final File f, final Dependency d)
...
+    return dependencies.
```
Comment (jakhani) [inline azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:190]: Can dependencies set itself be null?

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java [hunk 4]
```diff
@@ -73,7 +77,8 @@ public static DependencyFile getDependencyFile(final File projectFolder, final D
   public static void writeStartupDependencies(final File f,
       final Set<Dependency> dependencies) throws IOException {
     Map<String, Set<Dependency>> outputFormat = new HashMap<>();
-    outputFormat.put("dependencies", dependencies);
+    Set<Dependency> nonNullDependencies = filterNullFromDeps(dependencies);
+    outputFormat.put("dependencies", nonNullDependencies);
```
Comment (jakhani) [inline azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:81]: Can you please replace with outputFormat.put("dependencies", filterNullFromDeps(dependencies));

File: azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java [hunk 5]
```diff
@@ -91,8 +91,13 @@
       while (rs.next()) {
         // Columns are (starting at index 1): file_name, file_sha1, validation_status
         Dependency d = hashAndFileNameToDep.remove(rs.getString(1) + rs.getString(2));
-        FileValidationStatus v = FileValidationStatus.valueOf(rs.getInt(3));
-        depValidationStatuses.put(d, v);
+
+        // HashMap.remove will return null if the key is not found, hence check for it before
+        // adding to depValidationStatuses.
+        if (d != null) {
```
Comment (jakhani) [inline azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java:97]: It will be better to add contains check for key in hashAndFileNameToDep instead of null check over here. All of these should be inside that contains if case.

File: azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java [hunk 6]
```diff
@@ -272,6 +276,8 @@ private void updateValidationStatuses(final Set<? extends Dependency> untouchedD
         .keySet()
         .stream()
         .filter(d -> validationStatuses.get(d) == status)
+        .filter(Objects::nonNull) // Filter out null since the Set should contain Dependency
```
Comment (jakhani) [inline azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java:279]: nonNull should be above filter for validationStatuses

Comment (jakhani) [inline azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java:279]: nonNull should be above filter for validationStatuses

File: test/src/main/java/azkaban/test/executions/ThinArchiveTestUtils.java (hunk 1 above)
Comment (li-gsalia) [reply test/src/main/java/azkaban/test/executions/ThinArchiveTestUtils.java:138]: This is a method used only for Unit test and follows the convention in the file. The pre-existing test dependencies have names A, B and C and various methods in the file have the dependency names in the method name.

Since this is for unit test, all the dependencies are dummy and do not have real world significance.

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java (hunk 2 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:189]: Done.

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java (hunk 3 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:190]: The Set can be empty, but not null.

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java (hunk 4 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:81]: Changed.

File: azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java (hunk 6 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java:279]: Changed the order

File: azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java [hunk 7]
```diff
@@ -281,4 +287,5 @@ private void updateValidationStatuses(final Set<? extends Dependency> untouchedD
         .map(d -> String.format("Removed blacklisted file %s", d.getFileName()))
         .collect(Collectors.toSet());
   }
+
```
[the remark this answers is not in the document]
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java:290]: It was added accidentally, thanks for catching it. Removed.

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:189]: Removed the FQN.

File: azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java (hunk 5 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java:97]: Contains check will lead to 2 lookups in the hasMap for every dependency instead of 1 which is the current case. Let me know if you feel strongly about using contains, I can make the change.

File: azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java (hunk 6 above)
Comment (djaiswal83) [reply azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java:279]: @li-gsalia the order appears same. Did you push the change?

File: azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java (hunk 5 above)
Comment (djaiswal83) [reply azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java:97]: I agree with @li-gsalia about it.

Comment (abhisheknath2011) [reply azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java:97]: If we are expecting more key miss i.e. keys not present in hashAndFileNameToDep, then contains check looks more logical. But I also agree with @li-gsalia. So will leave it up to @li-gsalia to take a call on this.

File: azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java (hunk 4 above)
Comment (abhisheknath2011) [reply azkaban-common/src/main/java/azkaban/utils/ThinArchiveUtils.java:81]: looks like the changes are not published.

File: azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java (hunk 5 above)
Comment (jakhani) [reply azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java:97]: +1 with @abhisheknath2011

File: azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java (hunk 6 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java:278]: Pushed the change now, laptop had issues yesterday.

File: azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java (hunk 5 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/project/JdbcDependencyManager.java:97]: There should be no miss based on the current code. Added the null check since set.remove() can return null. This will future proof the code in case future code changes lead to a miss in set.

File: azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java (hunk 6 above)
Comment (li-gsalia) [reply azkaban-common/src/main/java/azkaban/project/ArchiveUnthinner.java:278]: Pushed the change now, had some laptop issues yesterday.

**Review by abhisheknath2011 — APPROVED** [review]
Thanks for identifying the issue and making the changes. Changes looks good to me.

**Comment by abagshaw** [comment]
@li-gsalia Thanks for fixing my bug 😄

An unsolicited 2 cents (maybe you already looked into this, but just in case...): I'd keep an eye out for issues like flows failing with `ClassNotFoundException` in projects that were hitting this `null` dependency issue. It might be the case that the root cause of the problem is on the Gradle side of things (the thinning step at build time that generates the `startup-dependencies.json` file) and some JAR is being thinned (removed from the zip) but failing to be added to the list of thinned dependencies (i.e. it's being added as `null` for some reason) which could result it it being missing from the classpath at runtime.

**Comment by li-gsalia** [comment]
@abagshaw it's good to see you being active on Azkaban! :) 

For the projects where this issue was seen, `startup-dependencies.json` does not have a `null` in it after the build completes. In fact, the behavior is not reproducible for the same `startup-dependencies.json` file. The null gets inserted during byte-ray processing of the dependencies. The new checks added will ensure `null` is not saved in the DB.

Thanks for pointing out the possibility of `null` being added during build time. Will keep an eye out for it.
